proposals/xxxx-super-call-enforcement.md

```diff
@@ -0,0 +1,200 @@
+
+```swift
+class C1 {
+  @requiresSuper 
+  func foo() {}
+}
+
+class C2: C1 {
+  override func foo() {} // error: method override is missing 'super.foo()' call
+}
+
+class C3: C1 {
+  override func foo() { // Okay
+    super.foo()
+  }
+}
+```
+
+## Detailed Design
+
+The `super` call must be made to the base method that is annotated with `@requiresSuper` and not to some other base method. For example:
+
+```swift
+class C1 {
+  @requiresSuper
+  func foo() {}
+  func foo(arg: Bool) {}
+  func foo(arg: Int = 3) {}
+}
+
+class C2: C1 {
+  override func foo() { super.foo(arg: true) } // error: method override is missing 'super.foo()' call
+}
+
+class C3: C1 {
+  override func foo() { super.foo(arg: 0) } // error: method override is missing 'super.foo()' call
+}
+
+class C4: C1 {
+  override func foo() { super.foo() } // Okay (since this is calling 'foo()' and not 'foo(arg: Int = 0)')
+}
+```
+
+A few more examples:
+
+```swift
+class C1 {
+  @requiresSuper
+  func foo(_ arg: Int) {}
+  func foo(_ arg: Bool) {}
+}
+
+class C2: C1 {
+  override func foo(_ arg: Int) { super.foo(true) } // error: method override is missing 'super.foo(_:)' call
+}
+
+class C3: C1 {
+  override func foo(_ arg: Int) { super.foo(0) } // Okay
+}
+```
+
+A message can be (optionally) specified on the `@requiresSuper` attribute to provide any additional information, which will be shown with the error message:
+
+```swift
+class C1 {
+  @requiresSuper("Call super as the final step in your implementation")
+  func foo() {}
+}
+
+class C2: C1 {
+  override func foo() {} // error: method override is missing 'super.foo()' call: Call super as the final step in your implementation
+}
+```
+
+Any overrides of a method annotated with `@requiresSuper` will also implicitly inherit the attribute, to make sure that all the overrides in a subclass chain call back to the `super` method. This will be suppressed if the override is in a `final` class:
+
+```swift
+class C1 {
+  @requiresSuper
+  func foo() {}
+}
+
+class C2: C1 {
+  // Implicitly inherits '@requiresSuper' from 'C1.foo'
+  override func foo() {
+    super.foo()
+    // Do some other stuff
+  }
+}
+
+final class C3: C1 {
+  // Does not implicitly inherit '@requiresSuper' from 'C1.foo', since 'C3.foo' cannot be overridden
+  override func foo {
+    super.foo()
+    // Do some other stuff
+  }
+}
+
+class C4: C2 {
+  override func foo() {} // error: method override is missing 'super.foo()' call
+}
+```
+
+Finally, a `@requiresSuper` attribute will also be implicitly added for imported ObjC methods which have been annotated with `objc_requires_super`. This enables warnings (instead of errors) for any method with such an attribute, such as `UIViewController.viewDidLoad()`:
```

Comment:
`UIViewController.viewDidLoad()` doesn't appear to have the `NS_REQUIRES_SUPER` attribute.

Examples from UIKit:

* [`UIView.updateConstraints()`](https://developer.apple.com/documentation/uikit/uiview/1622512-updateconstraints)
  > Call [super updateConstraints] as the final step in your implementation.

* [`NSObject.awakeFromNib()`](https://developer.apple.com/documentation/objectivec/nsobject/1402907-awakefromnib)
  > You must call the super implementation of awakeFromNib to give parent classes the opportunity to perform any additional initialization they require. Although the default implementation of this method does nothing, many UIKit classes provide non-empty implementations. You may call the super implementation at any point during your own awakeFromNib method.